### ISSUE TITLE
Is there a way to kill off GoRoutines?

### ISSUE THREAD

**Issue by boyter** [dbaggerman/cuba#1]
Been looking through and not noticing anything, but is there some way to add a bail out to all the GoRoutines spawned by cuba?

Im toying with a terminial UI interface running some search using it. Because I want the UI to be responsive I debounce the input BUT when a new one comes in I need to kill off any background GoRoutines so I can then spawn a new batch to do the next search.

Finish() appears to do something like this but waits for everything to clean up. Ideally id like a FinishIncomplete or something similar which blocks killing all the background GoRoutines and then returns. I have been looking though and I don't see anything that does do this, but I might be missing something.

I think it would be a case of modifying for for loop in `runWorker` such that it can bail out on command but not 100% sure on that. Would emptying the pool be the best way to achieve this or some sort of flag to bail out?

To explain further, what I am seeing is that the number of GoRoutines grows without bounds when running over large directories with many thousands of files and directories. I need to control this to avoid ulimit issues and the like.

**Comment by dbaggerman** [comment]
Yep, `Finish()` waits for all the work to complete - including anything additional the workers might add after the call to `Finish()`. I hadn't gotten around to adding a way to abort the process when I did the first round of work.

I could add a method to clear any pending items out of the queue, and prevent the workers adding more.

Actually aborting the active goroutines is trickier, since the go runtime doesn't let you kill goroutines directly. Probably the best I could do there is to provide a way to get a [`context.Context`](https://golang.org/pkg/context/#Context) into the worker, but it would still be up to the worker to check the context and shut itself down.

**Comment by boyter** [comment]
Yep clearing the queue seems to work with my tweaks. Assuming the producers outside stop as well that should clear up... might need to just block till the waitgroups finish. 

I added the following into `pool.go` which I think achieves the goal,

```
func (pool *Pool) EmptyPool() {
	pool.bucket.Empty()
	pool.wg.Wait()
}
```

**Comment by dbaggerman** [comment]
`bucket.Empty()` only checks if the bucket is empty, it doesn't remove anything currently inside. Also, the buckets weren't written to be threadsafe themselves -- the `Pool` has a mutex that needs to be locked when accessing the bucket.

Also, even if you empty the bucket anything the workers add back into the queue would still be processed. The `pool.closed` flag probably needs to be extended to allow for an aborted state to drop anything that gets pushed after the abort.

I'll have a look at it over the weekend.

**Comment by dbaggerman** [comment]
I've created a [branch](https://github.com/dbaggerman/cuba/tree/issues/1-abort) with a `pool.Abort()` method to stop accepting new work and wait for the goroutines to close down.

Could you try it out and see if it solves your problem?

**Comment by boyter** [comment]
Yep will do.

**Comment by dbaggerman** [comment]
@boyter, did you ever get around to trying this? If not, are you still planning to?

Otherwise, I'll just merge it as is.

**Comment by boyter** [comment]
@dbaggerman I have not as yet sorry. I got distracted working on some other stuff. Feel free to merge if you like because I won't be able to get to this till after October 31st I am sorry to say.

**Comment by boyter** [comment]
100% I will be looking at it again after that date though.

**Comment by dbaggerman** [comment]
No worries. I've merged that functionality as v0.3.0.

If you have any further feedback once you've had a chance to try it out, just raise new bugfix/enhancement issues.
